noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr

```diff
@@ -376,6 +395,81 @@ pub(crate) comptime fn create_message_discovery_call() -> Quoted {
     }
 }
 
+/// Injects an authwit verification check of the form:
+/// ```
+///   if (!from.eq(context.msg_sender())) {
+///         assert_current_call_valid_authwit::<N>(&mut context, from);
+///     } else {
+///         assert(authwit_nonce, "Invalid auhtwit nonce. When 'from' and 'msg_sender' are the same, authwit_nonce must be zero");
+///     }
+/// ```
+/// where `from` and `authwit_nonce` are the names of the parameters that are expected to be present in the function definition.
+/// This check is injected by the `#[authorize_once("from_arg_name", "nonce_arg_name")]`, which allows the user to define
+/// which parameters to use.
+pub(crate) comptime fn create_authorize_once_check(f: FunctionDefinition) -> Quoted {
+    let name = f.name();
+    let maybe_authorize_once_args = AUTHORIZE_ONCE_REGISTRY.get(f);
+    if maybe_authorize_once_args.is_none() {
+        panic(
+            f"Functions marked with #[authorize_once] must have the #[private] or #[public] attribute placed last",
+        );
```

Comment:
??

Reply:
Since it's a macro with params, we have to run it first to store them. Then we retrieve them in `#[private]` or `#[public]`. Unfortunate consequences of the fixed ordering of stuff.

Reply:
Sweet god. Can you leave a comment?